plugins/inputs/ethtool/ethtool_linux.go

```diff
@@ -143,41 +192,108 @@ func NewCommandEthtool() *CommandEthtool {
 }
 
 func (c *CommandEthtool) Init() error {
-	if c.ethtool != nil {
-		return nil
-	}
-
-	e, err := ethtoolLib.NewEthtool()
-	if err == nil {
-		c.ethtool = e
-	}
-
-	return err
+	return nil
 }
 
-func (c *CommandEthtool) DriverName(intf string) (string, error) {
-	return c.ethtool.DriverName(intf)
+func (c *CommandEthtool) DriverName(intf NamespacedInterface) (driver string, err error) {
+	return intf.Ethtool.DriverName(intf.Name)
 }
 
-func (c *CommandEthtool) Stats(intf string) (map[string]uint64, error) {
-	return c.ethtool.Stats(intf)
+func (c *CommandEthtool) Stats(intf NamespacedInterface) (stats map[string]uint64, err error) {
+	return intf.Ethtool.Stats(intf.Name)
 }
 
-func (c *CommandEthtool) Interfaces() ([]net.Interface, error) {
-	// Get the list of interfaces
+func (c *CommandEthtool) Interfaces() ([]NamespacedInterface, error) {
+	allInterfaces := make([]NamespacedInterface, 0)
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// Get the list of interfaces in the init namespace
+	initialNamespace, err := netns.Get()
+	if err != nil {
+		return nil, err
+	}
+	e, err := ethtoolLib.NewEthtool()
+	if err != nil {
+		return nil, err
+	}
 	interfaces, err := net.Interfaces()
 	if err != nil {
 		return nil, err
 	}
+	for _, iface := range interfaces {
+		allInterfaces = append(
+			allInterfaces,
+			NamespacedInterface{
+				Interface:     iface,
+				Namespace:     initialNamespace,
+				NamespaceName: "",
+				Ethtool:       e,
+			},
+		)
+	}
+
+	namespaces, err := os.ReadDir("/var/run/netns")
+	if err != nil {
+		c.Log.Warnf("Could not find namespace directory: %s", err)
+		return allInterfaces, nil
+	}
```

Comment:
Should the filtering of the interfaces happen before we hit this point? This way telegraf only reads this directory and attempts to change namespaces, if we have a list of namespaces to actually review.

Reply:
That's one option.

It would move even more of the logic out of the abstraction layer and into the implementation layer, which also makes the unit tests less useful because they're testing less of the runtime logic. Which is why I didn't go with that approach initially.

But it would be overall more performant to filter the interfaces and namespaces sooner, plus, in the namespace case, would reduce annoying log warning spam.

Reply:
Further testing revealed that this approach has very noticeable performance impact on my workload. Going to see what can be done to improve performance.